core/templates/dev/head/pages/collection_editor/editor_tab/CollectionLinearizerService.js

```diff
@@ -20,26 +20,34 @@
  */
 
 oppia.factory('CollectionLinearizerService', [
-  'CollectionUpdateService', 'SkillListObjectFactory',
-  function(CollectionUpdateService, SkillListObjectFactory) {
+  'CollectionUpdateService',
+  function(CollectionUpdateService) {
     var _getNextExplorationIds = function(collection, completedExpIds) {
-      var acquiredSkillList = completedExpIds.reduce(
-        function(skillList, explorationId) {
+      var userAcquiredSkillIds = completedExpIds.reduce(
+        function(skillIdsList, explorationId) {
           var collectionNode = collection.getCollectionNodeByExplorationId(
             explorationId);
-          skillList.addSkillsFromSkillList(
-            collectionNode.getAcquiredSkillList());
-          return skillList;
-        }, SkillListObjectFactory.create([]));
+          collectionNode.getAcquiredSkillIds().forEach(function(skillId) {
+            if (skillIdsList.indexOf(skillId) === -1) {
+              skillIdsList.push(skillId);
+            }
+          });
+          return skillIdsList;
+        }, []);
 
       // Pick all collection nodes whose prerequisite skills are satisified by
       // the currently acquired skills and which have not yet been completed.
       return collection.getExplorationIds().filter(function(explorationId) {
+        if (completedExpIds.indexOf(explorationId) !== -1) {
+          return false;
+        }
+
         var collectionNode = collection.getCollectionNodeByExplorationId(
           explorationId);
-        return completedExpIds.indexOf(explorationId) === -1 &&
-          acquiredSkillList.isSupersetOfSkillList(
-            collectionNode.getPrerequisiteSkillList());
+        return collectionNode.getPrerequisiteSkillIds().every(
```

Comment:
Eh, not sure. I think it'll lead to a proliferation of functions in the collection node object so I would lean against unless it's being used multiple times. I think .every() isn't too bad though.